### ISSUE TITLE
Unhandled exception in asynchronous task brings down database

### ISSUE THREAD

**Issue by mtornwall** [Starcounter/Starcounter#3032]
If an exception is thrown and left unhandled in an asynchronously executed task the whole database instance terminates. Shortest repro:

```
new DbSession().RunAsync(() => { throw new Exception(); });
```

The asynchronous versions of the methods in `Node`, `Http` etc all seem to suffer from this problem as well.

Running Starcounter v2.1.257.

**Comment by per-samuelsson** [comment]
@mtornwall 

Thanks for reporting this! We'll surely look into it ASAP.

Not sure who'll take this, @dan31?

**Comment by dan31** [comment]
Hi guys! @chrhol + @alemoi is a way to go. Both are on vacations, @chrhol is back next week in the end of it. Thanks for reporting!

**Comment by per-samuelsson** [comment]
> Hi guys! @chrhol + @alemoi is a way to go. Both are on vacations, @chrhol is back next week in the end of it. Thanks for reporting!

Sounds good enough. After reconsidering it, maybe P/High is borderline, since its quite easy to get around it, just not throwing any exceptions at this spot. At least thats why its not a showstopper.

But since its still somewhat serious, I guess that label is fine.

**Comment by mtornwall** [comment]
> Sounds good enough. After reconsidering it, maybe P/High is borderline, since its quite easy to get around it, just not throwing any exceptions at this spot. At least thats why its not a showstopper.

Agreed. A responsibly written asynchronous callback will likely do some kind of error handling on its own anyway (I'd like to think that's why it has taken me so long to get bitten by this one :-)).

I'm not sure what I think the appropriate behavior is, to be honest. Clearly it's no good to bring down the entire node. On the other hand, if exceptions occurring in callbacks are simply suppressed and logged it could result in semi-silent failures where some piece of code waiting on the callback never gets invoked. Neither scenario is particularly appealing.

**Comment by dan31** [comment]
@per-samuelsson, considering the amount of priors for those two happy guys when they are back, I'd say this one is "whenever possible" :) Luckily, we have a bunch of related issues there. E.g. we've recently addressed that part to optimise the download server. Also, considering that new scheduler (with async support and other nice stuff) is to be addressed in the upcoming months, finding some substantial solution for the existing one might be less meaningful.

**Comment by alemoi** [comment]
I thought it was fixed already, so that entry about exception is added to the server log.

**Comment by mtornwall** [comment]
@alemoi I can confirm that an entry is indeed written to the log.

However, I'm questioning whether it's reasonable for an error in an asynchronous task to bring down the entire database.

Here's something to compare with. If you run the following in a regular .NET console application:

    Task.Run(() => { throw new Exception(); });

... the program does *not* terminate.

**Comment by bigwad** [comment]
Hi, 
Just two cents.
Actually this Task's behavior has been changed recently for better async/await support. Here is [more details](http://blogs.msdn.com/b/pfxteam/archive/2011/09/28/task-exception-handling-in-net-4-5.aspx). You still can experience program crash in this case if specify [ThrowUnobservedTaskExceptions]( https://msdn.microsoft.com/en-us/library/jj160346(v=vs.110).aspx) option in app.config
On top of that DbSession.RunAsync() is more like ThreadPool.QueueUserWorkItem() than a Task.Run() since they both follow "fire and forget" pattern. And unhandled exception in ThreadPool.QueueUserWorkItem will definitely crash your program.

Anyway, here is obviously a lack of modern way of handling exceptions inside RunAsync() method.

**Comment by alemoi** [comment]
I agree that exception should not bring down the database. For me RunAsync is like any other code inside the main thread.

**Comment by Starcounter-Jack** [comment]
Of course, the behaviour is not reasonable. User code bringing down the database is never reasonable.

**Comment by alemoi** [comment]
I guess by "entire database" we are talking about hosting code process only. Is that right @mtornwall?

**Comment by mtornwall** [comment]
> I guess by "entire database" we are talking about hosting code process only. Is that right @mtornwall?

Sorry, I should've expressed myself more clearly. What I've *actually* observed is that the code host and the particular database instance to which it belonged terminated (as in: no longer listed as Started in the administrator).

**Comment by alemoi** [comment]
As far as I know, the database is shown to be running in Administrator, if and only if the codehost is up (it does not care about scdata process, which the actual database engine). @andwah @per-samuelsson  please correct me if I am wrong.

I was thinking a bit, and the problem with just logging and not terminating sccode is that developer might not notice that something died. Usually if most of functionality still runs - one might think that everything else is running fine. One example is a RunAsync call to cleanup some resources once in a while. If this functionality dies alone, no one notices this and application runs until something else breaks down, but then it will be hard to find the initial cause. It was the case with session cleanup, for example.

**Comment by andwah** [comment]
The database is shown "running" if there is a database and a host process running.

**Comment by per-samuelsson** [comment]
> As far as I know, the database is shown to be running in Administrator, if and only if the codehost is up (it does not care about scdata process, which the actual database engine). @andwah @per-samuelsson please correct me if I am wrong.

> The database is shown "running" if there is a database and a host process running.

In response to the above, if you ever need the details, the CLI will reveal them.

Do first

```
C:\Users\Per>staradmin start db
C:\Users\Per>staradmin list db
default
  Running         True
  Host process ID 1592
  Applications    0
```

and then

```
C:\Users\Per>staradmin stop host
C:\Users\Per>staradmin list db
default
  Running         True
  Host process ID n/a
  Applications    n/a
```

**Comment by dan31** [comment]
@chrhol, I just came across this one viewing multiple references pointing to it. I think there should be more issues that should cross-reference this one. If you are mostly done with the pnext track, can you look in this one? I don't remember whether we discussed it, actually. Though we might come up with some relevant fix, the problem might actually surface a (known) need to improve the scheduler. We can chat about that one at the office together.

**Comment by chrhol** [comment]
Posted this from #3329 to keep the discussion in one place.

> > The behavior I've come to expect is that when exceptions occur, they are allowed to unwind as much as possible of the stack up to the point where they would bring down the process. At that point the exception is caught and logged. I think the behavior should be similar here.

> That's very standard, yes. And I think we should rarely diverge from that pattern, and be especially careful we never catch exceptions that we don't actually handle, but instead goes ignored. That include catching exceptions and log them without any further reaction.

> WTBS, things are somewhat a little different in server models, and certainly so in our model, where we have to consider not only that the database / host keep running, but also that other apps in the same process keep doing that. One failing app should never be allowed to fuck it up for others.

> The right thing to do seem for me to be that the host would catch exceptions that slip out from any application, assure as much information as possible is logged (including context and inner exceptions) and then **terminate the application**.

This is a quite big change on how we handle unhandled exceptions in general in Starcounter. Can we even do this and determine that the state of the other apps are ok and not corrupted in any way? So we get some weird errors later that we have no idea where they originated from.

**Comment by chrhol** [comment]
Just to clarify: I meant the last bold part, terminate the application, as a resolution instead of terminating the codehost as the big change. I agree on the other parts.

**Comment by bigwad** [comment]
Did we consider using appdomains for applications?

**Comment by per-samuelsson** [comment]
> Did we consider using appdomains for applications?

We certainly did, many times. But always came to the conclusion that cons outweigh the pros. And we don't want to go in that direction now, since they are being moved away from in the new DNX stack / `coreclr`.

**Comment by alemoi** [comment]
I will make sure that we just log the exception and not re-throwing it when it comes to `ScheduleTask` API.

**Comment by bigwad** [comment]
If you're not going to rethrow then process will be functioning in unpredictable state. Is it your real intention?

I support the opinion of @per-samuelsson:
>The right thing to do seem for me to be that the host would catch exceptions that slip out from any application, assure as much information as possible is logged (including context and inner exceptions) and then terminate the application.

except that in general logging may make things worse, but that hardly applies to managed exceptions.

**Comment by alemoi** [comment]
> If you're not going to rethrow then process will be functioning in unpredictable state. Is it your real intention?

That's a good argument, as always from you. We didn't think deep now.
True, the waiting code for the task, expects a correct state after the task is finished. If exception is thrown in such task, then calling code might never now that it happened. This refers more to our general discussion on our current tasks implementation - one does not get a Task object to, later, control the state of this task.

**Comment by alemoi** [comment]
Here are some case variants regarding termination and not termination:
```
        async static void SomeTask() {
            Console.WriteLine("1");
            await Task.Delay(1000);

            throw new Exception("hahaha");
        }

        // Terminates.
        static void Main(string[] args) {
            Console.WriteLine("0");
            Thread t = new Thread(() => { SomeTask(); });
            t.Start();
            Console.WriteLine("2");
            Thread.Sleep(5555555);
        }

        // Terminates.
        static void Main2(string[] args) {
            Console.WriteLine("0");
            var t = Task.Run(() => {
                Console.WriteLine("1");
                throw new Exception("hahaha");
            });
            t.Wait();
            Console.WriteLine("2");
            Thread.Sleep(5555555);
        }

        // Terminates.
        static void Main3(string[] args) {
            Console.WriteLine("0");
            Thread t = new Thread(() => {
                Console.WriteLine("1");
                throw new Exception("hahaha");
            });
            t.Start();
            Console.WriteLine("2");
            Thread.Sleep(5555555);
        }

        // Does NOT terminate.
        static void Main4(string[] args) {
            Console.WriteLine("0");
            Task.Run(() => {
                Console.WriteLine("1");
                throw new Exception("hahaha");
            });
            Console.WriteLine("2");
            Thread.Sleep(5555555);
        }
```

**Comment by dan31** [comment]
@alemoi how does it look like in case of Starcounter's ```ScheduleTask``` (sync/async)?
Regarding the latter, I guess since the task is executed on a different thread, the exception isn't propagated onto the current thread. Do you find an acceptable explanation for the latter behavior? Is this behavior recommended to avoid for developers?

**Comment by alemoi** [comment]
The task is now always scheduled but I can't find the logging part. I guess that was missed during big restructuring in there. I will introduce it back and test it.

**Comment by dan31** [comment]
The tas is always scheduled - what does this mean? I think that, following the least surprise principle (and @bigwad in this respect) you could simply mimic the .net behaviour. Which is - throwing exception except for the last case where it is intercepted. Hence if we answer why it is intercepted (and prove it is no good) our job here is done, right? So... Why?

**Comment by bigwad** [comment]
> So... Why?

Here is combination of reasons.
First, prior to .Net 4.5, if you don't wait for a task, you'll get this exception when task will be gc'ed on gc thread and it will crash the process by default. So if you place GC.Collect(), it should crash.
But, starting from .Net 4.5 it won't crash by default to make async/await programming more convenient as I already mentioned in this thread: https://github.com/Starcounter/Starcounter/issues/3032#issuecomment-154440112

**Comment by eriksunsol** [comment]
I read this thread with interest. I must admit I had some trouble following the reasoning in the last couple of comments, but here's just my two cents:

Everything else aside, I think it is totally unacceptable that the server goes down when there is an unhandled exception. Anywhere. Unless it is caused by some internal error in Starcounter from which it is impossible to recover. Excessive logging and performance degradation may be bad, but letting the server go down is worse.

For comparison, application containers for JaveEE (or at least jBoss which I have experience with) are darn near impossible to crash. The only way I have had jBoss crash is from memory exhaustion, which I think is acceptable as it is really difficult to work around gracefully. I would expect the same from Starcounter.

As I understand it this problem is currently exposed when running Starcounter scheduled tasks, and yes, it's probably good practice to responsibly deal with exceptions in scheduled code, but good practice cannot always be enforced unfortunately.

So please, whatever you decide to do about this, never ever let the server crash!

If your are concerned that problems in application code may be hidden from administrators and developers by silently just logging these errors, why not e.g. require that a callback is registered somewhere to react on these errors. Then application developers can at least make a active choice whether to ignore, log, crash or whatever is appropriate.

**Comment by bigwad** [comment]
When we're talking that unhandled exception brings down the server (or database) it's not actually what takes place here. 

Logically and physically, it's the user application that goes down. Bringing down the app is a safety measure and all your data is safe and will be in place on app restart. As for industry standards: I'm not aware about how JBoss deals with this in details (will take a look), but IIS for example brings down ASP.NET app in such case [link1](https://support.microsoft.com/en-us/kb/911816), [link2](https://blogs.msdn.microsoft.com/tess/2008/08/19/questions-on-application-domains-application-pools-and-unhandled-exceptions/)

Unfortunately, we don't provide proper isolation between apps as well as between apps and starcounter hosting code. Thus bringing down one app means the same for all. As @per-samuelsson noted [here](https://github.com/Starcounter/Starcounter/issues/3032#issuecomment-178497710) we're moving to better isolation.

Sure, if we go with bringing an app down, we need better app management regarding notifications, automatic app restart etc.

**Comment by per-samuelsson** [comment]
> When we're talking that unhandled exception brings down the server (or database) it's not actually what takes place here.

Well, in practice it is. It's nothing we do intentionally, because we never recover the host and neither those applications running as fellows to the application crashing. But the net effect for our users is that the experience their server is dead, and that's what counts.

> Everything else aside, I think it is totally unacceptable that the server goes down when there is an unhandled exception. Anywhere.

Totally agree.

**Comment by chrhol** [comment]
I'm not really sure why this issue is assigned to me or if we have found on an actual way forward (other than agreeing that this needs to be "fixed" somehow). 

@dan31: is this something we should change in 2.x, and if so who should do it? :smile:
